phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala

```diff
@@ -0,0 +1,371 @@
+
+  private var state: State = WaitingForServerToStart(Nil)
+  private val connectionClosed = Promise[Unit]()
+
+  private[this] val fragmentsBuf = new StringBuilder
+
+  private val promise = Promise[Unit]()
+
+  def future: Future[Unit] = promise.future
+
+  private def loadMgr(): WebsocketManager = {
+    val loader =
+      if (jettyClassLoader != null) jettyClassLoader
+      else getClass().getClassLoader()
+
+    val clazz = loader.loadClass(
+        "org.scalajs.jsenv.phantomjs.JettyWebsocketManager")
+
+    val ctors = clazz.getConstructors()
+    assert(ctors.length == 1, "JettyWebsocketManager may only have one ctor")
+
+    val listener = new WebsocketListener {
+      def onRunning(): Unit = onServerRunning()
+      def onOpen(): Unit = onConnectionOpened()
+      def onClose(): Unit = onConnectionClosed()
+      def onMessage(msg: String): Unit = receiveFrag(msg)
+      def log(msg: String): Unit = logger.debug(s"PhantomJS WS Jetty: $msg")
+    }
+
+    val mgr = ctors.head.newInstance(listener)
+
+    mgr.asInstanceOf[WebsocketManager]
+  }
+
+  private val mgr: WebsocketManager = loadMgr()
+
+  // Constructor
+  mgr.start()
+
+  private def onServerRunning(): Unit = synchronized {
+    state match {
+      case WaitingForServerToStart(sendQueue) =>
+        state = AwaitingConnection(sendQueue)
+        val comSetup = makeComSetupFile(mgr.localPort)
+        val underlyingRun = startRun(comSetup)
+
+        underlyingRun.future.onComplete { result =>
+          underlyingRun.close()
+          onUnderlyingRunTerminated(result)
+        }
+
+      case Closing =>
+        // Ignore
+
+      case AwaitingConnection(_) | Connected | NotConnectedAnymore =>
+        throw new IllegalStateException(
+            s"Illegal state in onServerRunning: $state")
+    }
+  }
+
+  private def onConnectionOpened(): Unit = synchronized {
+    state match {
+      case AwaitingConnection(sendQueue) =>
+        sendQueue.reverse.foreach(sendNow)
+        state = Connected
+
+      case Closing =>
+        mgr.closeConnection()
+
+      case WaitingForServerToStart(_) | Connected | NotConnectedAnymore =>
+        throw new IllegalStateException(
+            s"Illegal state in onConnectionOpened: $state")
+    }
+  }
+
+  private def onConnectionClosed(): Unit = synchronized {
+    connectionClosed.success(())
+
+    state match {
+      case Connected =>
+        close()
+
+      case Closing =>
+        // Ignore
+
+      case WaitingForServerToStart(_) | AwaitingConnection(_) | NotConnectedAnymore =>
+        throw new IllegalStateException(
+            s"Illegal state in onConnectionClosed: $state")
+    }
+  }
+
+  private def onUnderlyingRunTerminated(result: Try[Unit]): Unit = synchronized {
+    state match {
+      case Connected | Closing =>
+        // Wait until the connection is closed before completing the promise
+        connectionClosed.future.foreach { _ =>
+          promise.tryComplete(result)
+        }
```

Comment:
Couldn't you do this in the `onConnectionClosed` callback instead of having an additional promise?

Reply:
No. There is no guarantee about the order in which `onConnectionClosed()` and `onUnderlyingRunTerminated()` will be executed. So we need this to happen when both have been called, with the `result` received here. The additional promise is the easiest way to do it, although of course I could also use another field with a few states, and the second one that is called triggers the `promise.tryComplete`. Would you prefer that?